### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by vgarvardt**
Things that are missing and will be added in this PR:

- [x] Readme with examples
- [x] Memory backend
- [x] Unit tests
- [x] Code coverage report

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+language: php
+
+php:
+  - '5.6'
+  - '7.0'
+
+install:
+  - composer install -n
```
Comment (boekkooi-fresh) [inline .travis.yml]: Please also test with lowest dependencies see https://evertpot.com/testing-composer-prefer-lowest/

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
+language: php
+
+php:
+  - '5.6'
+  - '7.0'
```
Comment (boekkooi-fresh) [inline .travis.yml]: Please also test PHP 7.1

File: src/Bucket/MetricOperation.php [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+     */
+    public function __construct(array $operations = [])
+    {
+        parent::__construct(static::LENGTH);
+
+        // ensure that operations is not bigger than allowed
+        array_splice($operations, static::LENGTH);
+
+        foreach ($operations as $key => $value) {
+            $this->offsetSet($key, $value);
```
Comment (boekkooi-fresh) [inline src/Bucket/MetricOperation.php]: Since array in PHP can be a hash this may fail in a nice way. Maybe add a unit test with `['a'=> 'a', 'b' => 'b', 'c' => 'c']` as the $operations?

File: src/Bucket/MetricOperation.php [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+     * MetricOperation constructor.
+     *
+     * @param array $operations
+     */
+    public function __construct(array $operations = [])
+    {
+        parent::__construct(static::LENGTH);
+
+        // ensure that operations is not bigger than allowed
+        array_splice($operations, static::LENGTH);
```
Comment (boekkooi-fresh) [inline src/Bucket/MetricOperation.php]: I think you forgot to set the offset to 0 aka `array_splice($operations, 0, static::LENGTH);`

File: src/Bucket.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace HelloFresh\Stats;
+
+
+interface Bucket
```
Comment (boekkooi-fresh) [inline src/Bucket.php]: This interface and it's methods could do with some more docs.

File: src/Bucket/HTTPRequest.php [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+     *
+     * @param string $section
+     * @param RequestInterface $request
+     * @param bool $success
+     * @param HTTPMetricAlterCallback|null $callback
+     */
+    public function __construct($section, RequestInterface $request, $success, HTTPMetricAlterCallback $callback = null)
+    {
+        $this->request = $request;
+        $this->httpMetricAlterCallback = $callback;
```
Comment (boekkooi-fresh) [inline src/Bucket/HTTPRequest.php]: Why do we need to store `request` and `httpMetricAlterCallback`? Why not simply add them as args to `buildMetricOperation` and store that output?

File: src/Bucket/HTTPRequest.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+                $operation[$partsFilled] = $fragment;
+                $partsFilled++;
+                if ($partsFilled >= $operation::LENGTH) {
+                    break;
+                }
+            }
+        }
+
+        if (null != $this->httpMetricAlterCallback) {
```
Comment (boekkooi-fresh) [inline src/Bucket/HTTPRequest.php]: Please use `!==` :smile:

File: src/Bucket/HTTPRequest.php [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+        parent::__construct($section, $this->buildMetricOperation(), $success);
+    }
+
+    /**
+     * @return MetricOperation
+     */
+    public function buildMetricOperation()
+    {
+        $operation = new MetricOperation(strtolower($this->request->getMethod()));
+        if ($this->request->getUri()->getPath() != '/') {
```
Comment (boekkooi-fresh) [inline src/Bucket/HTTPRequest.php]: Please use `!==`

File: src/Bucket/HTTPRequest.php [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+            $partsFilled = 1;
+            foreach (explode('/', $this->request->getUri()->getPath()) as $fragment) {
+                if ($fragment == '') {
```
Comment (boekkooi-fresh) [inline src/Bucket/HTTPRequest.php]: Please use `===`

File: src/Client/Log.php [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+namespace HelloFresh\Stats\Client;
+
+
+use HelloFresh\Stats\Client;
+use HelloFresh\Stats\Incrementer;
+use HelloFresh\Stats\State;
+use HelloFresh\Stats\Timer;
+use Psr\Log\LoggerInterface;
+
+class Log extends Base implements Client
```
Comment (boekkooi-fresh) [inline src/Client/Log.php]: I think this naming maybe a little off this looks more like a factory class. Then again I have no clue what this is for :smile:

File: src/Client/Log.php [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    {
+        return new Timer\Memory();
+    }
+
+    /**
+     * @inheritdoc
+     */
+    protected function getIncrementer()
+    {
+        return new Incrementer\Log($this->logger);
```
Comment (boekkooi-fresh) [inline src/Client/Log.php]: Should we not store the `Incrementer\Log` instance for later use? (same question for [line 49](https://github.com/hellofresh/stats-php/pull/1/files#diff-fb64cec0d1d5379aeae9d01b5fee5c4dR49) and the [NoOp](https://github.com/hellofresh/stats-php/pull/1/files#diff-ef31bdea77aabeb99a212c370f814750) version)

File: src/Client/StatsD.php [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        $url = parse_url($dsn);
+
+        $params = parse_str(empty($url['query']) ? '' : $url['query']);
+        $options = [
+            'host' => $url['host'],
+            'port' => $url['port'],
+            'namespace' => empty($params['ns']) ? '' : $params['ns'],
+            'timeout' => empty($params['timeout']) ? null : (float)$params['timeout'],
+            'throwConnectionExceptions' => empty($params['error']) ? true : (bool)$params['error'],
```
Comment (boekkooi-fresh) [inline src/Client/StatsD.php]: Instead of all the empty check you could do:
```PHP
$urlInfo = parse_url($dsn);
$options = array_merge(
['ns' => '', 'timeout' => null, 'error' => true],
$urlInfo
);
```

File: src/Client/StatsD.php [hunk 13]
```diff
@@ -0,0 +1,66 @@
...
+    protected $client;
+
+    /**
+     * StatsD constructor.
+     *
+     * @param string $dsn statsd connection dsn
+     */
+    public function __construct($dsn)
+    {
+        $url = parse_url($dsn);
```
Comment (boekkooi-fresh) [inline src/Client/StatsD.php]: [`parse_url`](http://php.net/parse_url) can return `NULL` you probably want to check for this.

File: src/Factory.php [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+     *
+     * @param string $dsn
+     * @param LoggerInterface $logger
+     *
+     * @return Client
+     * @throws \Exception
+     */
+    public static function build($dsn, LoggerInterface $logger)
+    {
+        $url = parse_url($dsn);
```
Comment (boekkooi-fresh) [inline src/Factory.php]: Please use `parse_url($dsn, PHP_URL_SCHEME);`

File: .travis.yml (hunk 2 above)
Comment (vgarvardt) [reply .travis.yml]: I did, but had to remove it because of `PHPUnit\Framework\Exception: PHP Fatal error:  strict_types declaration must be the very first statement in the script in /home/travis/.phpenv/versions/7.1.10/bin/phpunit on line 2` - https://travis-ci.org/hellofresh/stats-php/jobs/289517498

Comment (boekkooi-fresh) [reply .travis.yml]: I think you can solve this by using your own version of PHPUnit not the one from travis since it is out of date (aka PHPUnit 6.4.1 [fixes](https://github.com/sebastianbergmann/phpunit/issues/2791) this issue)

Comment (vgarvardt) [reply .travis.yml]: `phpunit/phpunit 6.4.3 requires php ^7.0` and we need to support 5.6

Comment (boekkooi-fresh) [reply .travis.yml]: I know and you are using 6.4.0 since that is what is installed on travis (see [line 11](https://github.com/hellofresh/stats-php/pull/1/files#diff-354f30a63fb0907d4ad57269548329e3R11)) you probably want to call your installed version `vendor/bin/phpunit` instead. 

To my knowledge this bug was introduced in phpunit 6.4.0 so 4.8.0 should work fine

File: src/Bucket/MetricOperation.php (hunk 4 above)
Comment (vgarvardt) [reply src/Bucket/MetricOperation.php]:
```php
$a = ['bar', 'baz', 'qaz', 'qux'];
$r = array_splice($a, 0, 3);
var_dump($r);
var_dump($a);
```

results in:

```
array(3) {
  [0]=>
  string(3) "bar"
  [1]=>
  string(3) "baz"
  [2]=>
  string(3) "qaz"
}
array(1) {
  [0]=>
  string(3) "qux"
}
``

And I do not understand why, although according to function description you're correct.

File: .travis.yml (hunk 2 above)
Comment (vgarvardt) [reply .travis.yml]: Works fine with `5.7.23` and `5.7.0`, so added 7.1 back.

File: tests/Bucket/HTTPRequestTest.php [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+            ->will($this->returnValue('GET'));
+        $request->expects($this->atLeastOnce())
+            ->method('getUri')
+            ->will($this->returnValue($uri));
+
+        /** @var \PHPUnit_Framework_MockObject_MockObject|\HelloFresh\Stats\HTTPMetricAlterCallback $callback */
+        $callback = $this->getMockBuilder('\HelloFresh\Stats\HTTPMetricAlterCallback')->getMock();
+
+        $callback->expects($this->once())
+            ->method('__invoke')
```
Comment (boekkooi-fresh) [inline tests/Bucket/HTTPRequestTest.php]: Should we not care about the arguments here?

File: tests/Bucket/MetricOperationTest.php [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+                ['foo', 'bar', 'baz'],
+            ],
+            [
+                ['foo', 'bar', 'baz'],
+                ['foo', 'bar', 'baz', 'qux'],
+            ],
+            [
+                ['foo', 'bar', 'baz'],
+                ['one' => 'foo', 'two' => 'bar', 'three' => 'baz', 'four' => 'qux'],
+            ],
```
Comment (boekkooi-fresh) [inline tests/Bucket/MetricOperationTest.php]: Can we add a test case for none string types here? like `new StdClass()` and `null` values?

File: tests/Client/LogTest.php [hunk 17]
```diff
@@ -0,0 +1,219 @@
...
+use HelloFresh\Stats\Bucket\Plain;
+use HelloFresh\Stats\Incrementer;
+use HelloFresh\Stats\State;
+use PHPUnit\Framework\TestCase;
+
+class LogTest extends TestCase
+{
+    public function setUp()
+    {
+        mt_srand(time());
```
Comment (boekkooi-fresh) [inline tests/Client/LogTest.php]: Why do we need this?

File: tests/Client/LogTest.php [hunk 18]
```diff
@@ -0,0 +1,219 @@
...
+        $logger = $this->getMockBuilder('\Psr\Log\LoggerInterface')->getMock();
+
+        $statsClient = new Log($logger);
+        $this->assertInstanceOf(Stats\Timer\Log::class, $statsClient->buildTimer());
+
+        $reflection = new \ReflectionClass($statsClient);
+
+        $methodIncrementer = $reflection->getMethod('getIncrementer');
+        $methodIncrementer->setAccessible(true);
+        $this->assertInstanceOf(Incrementer\Log::class, $methodIncrementer->invoke($statsClient));
```
Comment (boekkooi-fresh) [inline tests/Client/LogTest.php]: :-1: You should not test none public methods. In this case you could create a manual mock that makes the methods public but I would advice against this since this method should be tested based on other units taht call it internally. (same go's for all other reflection calls)

File: src/Bucket/Plain.php [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    /**
+     * @param string $metric
+     *
+     * @return string
+     */
+    public static function sanitizeMetricName($metric)
+    {
+        if ($metric == "") {
```
Comment (rafaeljesus) [inline src/Bucket/Plain.php]: `===`

File: src/Bucket/HTTPRequest.php [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+        parent::__construct($section, $this->buildMetricOperation($request, $callback), $success);
+    }
+
+    /**
+     * @param RequestInterface $request
+     * @param HTTPMetricAlterCallback|null $callback
+     *
+     * @return MetricOperation
+     */
+    public function buildMetricOperation(RequestInterface $request, HTTPMetricAlterCallback $callback = null)
```
[the remark this answers is not in the document]
Comment (vgarvardt) [reply src/Bucket/HTTPRequest.php]: Based on https://github.com/hellofresh/stats-go/blob/master/bucket/bucket_http_request.go#L32 usage experience - this was initially private, but had to make it public to allow external usage, in some cases it is required to build a metric from request. So in PHP implementation made this public from the beginning. But will change it to `public static` as this method has nothing to do with the object state.

File: src/Timer/Log.php [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+     */
+    public function __construct(LoggerInterface $logger)
+    {
+        $this->logger = $logger;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function start()
```
[the remark this answers is not in the document]
Comment (vgarvardt) [reply src/Timer/Log.php]: Yes, you can reuse timer instance.

File: src/Timer/Log.php [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+    {
+        $this->startedAt = microtime(true);
+        return $this;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function finish($metric)
```
[the remark this answers is not in the document]
Comment (vgarvardt) [reply src/Timer/Log.php]: Yes, it is useful in some cases - finish get the metric as parameter, so you can start once and finish several times in the required checkpoints.

File: tests/Bucket/HTTPRequestTest.php [hunk 23]
```diff
@@ -0,0 +1,78 @@
...
+     * @param string $method
+     * @param string $path
+     * @param string $section
+     * @param string $metric
+     */
+    public function testHTTPRequestTest($method, $path, $section, $metric)
+    {
+        $uri = $this->getMockBuilder('\Psr\Http\Message\UriInterface')->getMock();
+
+        $uri->expects($this->atLeastOnce())
```
[the remark this answers is not in the document]
Comment (vgarvardt) [reply tests/Bucket/HTTPRequestTest.php]: Well, this is more about implementation, I care how the metric is being built, what parts of the request it uses.

File: tests/Bucket/HTTPRequestTest.php [hunk 24]
```diff
@@ -0,0 +1,78 @@
...
+            ->method('getPath')
+            ->will($this->returnValue($path));
+
+        /** @var \PHPUnit_Framework_MockObject_MockObject|\Psr\Http\Message\RequestInterface $request */
+        $request = $this->getMockBuilder('\Psr\Http\Message\RequestInterface')->getMock();
+
+        $request->expects($this->once())
```
[the remark this answers is not in the document]
Comment (vgarvardt) [reply tests/Bucket/HTTPRequestTest.php]: Well, this is more about implementation, I care how the metric is being built, what parts of the request it uses.

File: tests/Bucket/HTTPRequestTest.php [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+            ->method('getMethod')
+            ->will($this->returnValue($method));
+        $request->expects($this->atLeastOnce())
```
[the remark this answers is not in the document]
Comment (vgarvardt) [reply tests/Bucket/HTTPRequestTest.php]: Well, this is more about implementation, I care how the metric is being built, what parts of the request it uses.

File: tests/Client/LogTest.php (hunk 17 above)
Comment (vgarvardt) [reply tests/Client/LogTest.php]: I use `mt_rand` to generate test values, so need to seed it first

File: src/Client.php [hunk 26]
```diff
@@ -0,0 +1,92 @@
...
+     * @return self
+     */
+    public function trackState($section, MetricOperation $operation, $value);
+
+    /**
+     * @param HTTPMetricAlterCallback $callback
+     *
+     * @return self
+     */
+    public function setHTTPMetricAlterCallback(HTTPMetricAlterCallback $callback);
```
[the remark this answers is not in the document]
Comment (vgarvardt) [reply src/Client.php]: Unfortunately yes, this must be part of the interface - based on `go` implementation usage experience.

**Comment by stefvaessen** [comment]
+1

File: src/Timer/Log.php (hunk 22 above)
[the remark this answers is not in the document]
Comment (boekkooi-fresh) [reply src/Timer/Log.php]: Maybe the naming is a little off here then maybe `time` or `elapsed` would be better.

**Comment by boekkooi-fresh** [comment]
:+1: It would be since to have a example/demo of the library or a actual integration test
